source/service.go

```diff
@@ -123,37 +115,28 @@ func (sc *serviceSource) Endpoints() ([]*endpoint.Endpoint, error) {
 	return endpoints, nil
 }
 
-func (sc *serviceSource) endpointsFromHeadless(svc *v1.Service) ([]*endpoint.Endpoint, error) {
+func (sc *serviceSource) extractHeadlessEndpoint(svc *v1.Service, hostname string) []*endpoint.Endpoint {
 
 	var endpoints []*endpoint.Endpoint
 
-	// Check if a Headless Service definition and we need to generate the hostnames differently
-	headlessDomain, ok := svc.Annotations[headlessDomainAnnotationKey]
-
-	// We require some domain to be set, maybe should also use the fqdn-template?
-	if ok && headlessDomain != "" {
-
-		// Get all the Pods
-		if pods, err := sc.client.CoreV1().Pods(svc.Namespace).List(metav1.ListOptions{LabelSelector: labels.Set(svc.Spec.Selector).AsSelectorPreValidated().String()}); err != nil {
-			log.Errorf("List Pods of service[%s] error:%v", svc.GetName(), err)
-		} else {
-			for _, v := range pods.Items {
-
-				log.Debugf("Generating matching endpoint %s%s with HostIP %s", v.Spec.Hostname+headlessDomain, v.GetName(), v.Status.HostIP)
-				// To reduce traffice on the DNS API only add record for running Pods. Good Idea?
-				if v.Status.Phase == v1.PodRunning {
-					endpoints = append(endpoints, endpoint.NewEndpoint(v.Spec.Hostname+headlessDomain, v.Status.HostIP, endpoint.RecordTypeA))
-				} else {
-					log.Debugf("Pod %s is not in running phase", v.Spec.Hostname)
-				}
+	// Get all the Pods
+	if pods, err := sc.client.CoreV1().Pods(svc.Namespace).List(metav1.ListOptions{LabelSelector: labels.Set(svc.Spec.Selector).AsSelectorPreValidated().String()}); err != nil {
```

Comment:
In the spirit of the original request around convention and suggestion. I personally have not see this style of expression used much, where the `else` is referencing the result of the assignment in the guarded `if` statement and it took me a few seconds to figure out where that was even coming from.  

Personally I find it more clear to simply make the assignment and have a small `err != nil` check that short circuits which eventually you get used to reading all over the place. 

That's a completely arbitrary assessment though and may not be true to those who have worked and looked at larger code bases.

Reply:
Sounds good. I will take a look shortly.

Reply:
You mean something like this?

```go
pods, err := sc.client.CoreV1().Pods(svc.Namespace).List(metav1.ListOptions{LabelSelector: labels.Set(svc.Spec.Selector).AsSelectorPreValidated().String()})
	// Get all the Pods
	if err != nil {
		log.Errorf("List Pods of service[%s] error:%v", svc.GetName(), err)
	} else {}
```

Reply:
You can have an empty return to short circuit: https://play.golang.org/p/jpRxJ19fl5

Ex.
```
pods, err := sc.client.CoreV1().Pods(svc.Namespace).List(metav1.ListOptions{LabelSelector: labels.Set(svc.Spec.Selector).AsSelectorPreValidated().String()}); 

if err != nil {
 	log.Errorf("List Pods of service[%s] error:%v", svc.GetName(), err)
 	return 
} 

for _, v := range pods.Items {
	headlessDomain := v.Spec.Hostname + "." + hostname
	log.Debugf("Generating matching endpoint %s with HostIP %s", headlessDomain, v.Status.HostIP)
	// To reduce traffice on the DNS API only add record for running Pods. Good Idea?
	if v.Status.Phase == v1.PodRunning {
		endpoints = append(endpoints, endpoint.NewEndpoint(headlessDomain, v.Status.HostIP, endpoint.RecordTypeA))
	} else {
		log.Debugf("Pod %s is not in running phase", v.Spec.Hostname)
	}
}
```

Reply:
But the function should return the Endpoints no? Or should we redesign that part?